### PR TITLE
html renderers in discrete

### PR DESCRIPTION
**Pull request by dellaert**
This PR adds html renderers for all discrete types for which markdown existed, and additionally for `DiscreteValues`. Importantly, it also 
- moves Names to DiscreteKeys
- makes the free version `cartesianProduct` a static method of `Assignment`
- adds a specialization of that in `DiscreteValues`
- markdown in `DiscreteConditional` is significantly re-factored to avoid copy/paste

### CODE REVIEW

File: gtsam/discrete/tests/testDiscreteValues.cpp [hunk 1]
```diff
@@ -0,0 +1,76 @@
...
+  values[12] = 1;  // A
+  values[5] = 0;   // B
+  string expected =
+      "<div>\n"
+      "<table class='DiscreteValues'>\n"
+      "  <thead>\n"
+      "    <tr><th>Variable</th><th>value</th></tr>\n"
+      "  </thead>\n"
+      "  <tbody>\n"
+      "    <tr><th>B</th><td>'-</td></tr>\n"
```
Comment (ProfFan) [inline gtsam/discrete/tests/testDiscreteValues.cpp:60]: Should be `" <tr><th>B</th><td>-</td></tr>\n"`

File: gtsam/discrete/tests/testDiscreteValues.cpp [hunk 2]
```diff
@@ -0,0 +1,76 @@
...
+
+/* ************************************************************************* */
+// Check html representation with a value formatter.
+TEST(DiscreteValues, htmlWithValueFormatter) {
+  DiscreteValues values;
+  values[12] = 1;  // A
+  values[5] = 0;   // B
+  string expected =
+      "<div>\n"
+      "<table class='DiscreteValues'>\n"
```
Comment (ProfFan) [inline gtsam/discrete/tests/testDiscreteValues.cpp:55]: Should be double quotes escaped

File: gtsam/discrete/DiscreteBayesTree.cpp [hunk 3]
```diff
@@ -72,5 +72,23 @@ namespace gtsam {
     return ss.str();
   }
 
+  /* **************************************************************************/
+  std::string DiscreteBayesTree::html(
```
Comment (ProfFan) [inline gtsam/discrete/DiscreteBayesTree.cpp:76]: This is NOT tested

File: gtsam/discrete/DiscreteValues.cpp [hunk 4]
```diff
@@ -0,0 +1,97 @@
...
+  // Print out header row.
+  ss << "    <tr><th>Variable</th><th>value</th></tr>\n";
+
+  // Finish header and start body.
+  ss << "  </thead>\n  <tbody>\n";
+
+  // Print out all rows.
+  for (const auto& kv : *this) {
+    ss << "    <tr>";
+    ss << "<th>" << keyFormatter(kv.first) << "</th><td>\'"
```
Comment (ProfFan) [inline gtsam/discrete/DiscreteValues.cpp:79]:
```suggestion
    ss << "<th>" << keyFormatter(kv.first) << "</th><td>"
```

File: gtsam/discrete/DiscreteValues.cpp [hunk 5]
```diff
@@ -0,0 +1,97 @@
...
+
+  return ss.str();
+}
+
+string DiscreteValues::html(const KeyFormatter& keyFormatter,
+                            const Names& names) const {
+  stringstream ss;
+
+  // Print out preamble.
+  ss << "<div>\n<table class=\'DiscreteValues\'>\n  <thead>\n";
```
Comment (ProfFan) [inline gtsam/discrete/DiscreteValues.cpp:68]:
```suggestion
  ss << "<div>\n<table class=\"DiscreteValues\">\n  <thead>\n";
```

**Review by ProfFan — APPROVED** [review]
Approving with some comments to unblock

File: gtsam/discrete/tests/testDiscreteValues.cpp (hunk 1 above)
Comment (dellaert) [reply gtsam/discrete/tests/testDiscreteValues.cpp:60]: Done

File: gtsam/discrete/tests/testDiscreteValues.cpp (hunk 2 above)
Comment (dellaert) [reply gtsam/discrete/tests/testDiscreteValues.cpp:55]: I was wondering. I did some searching and [this](https://docs.microsoft.com/en-us/cpp/c-language/c-string-literals?view=msvc-170) says  "The single quotation mark (') can be represented without an escape sequence."

Comment (dellaert) [reply gtsam/discrete/tests/testDiscreteValues.cpp:55]: Or are yuo saying html needs double quotes?

Comment (dellaert) [reply gtsam/discrete/tests/testDiscreteValues.cpp:55]: I think single is fine: https://www.w3.org/TR/html4/intro/sgmltut.html#h-3.2.2

File: gtsam/discrete/DiscreteBayesTree.cpp (hunk 3 above)
Comment (dellaert) [reply gtsam/discrete/DiscreteBayesTree.cpp:76]: You are right.

File: gtsam/discrete/DiscreteValues.cpp (hunk 5 above)
Comment (dellaert) [reply gtsam/discrete/DiscreteValues.cpp:68]: I will just make them single quotes.

**Pull request merged**